File: Source/OldCode/Archive/2005 TVA Code Library/Version/CS/TVA.Core_CS/Net/Ftp/DataStream.cs
using System.Diagnostics;
using System.Linq;
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.IO;
using System.Net.Sockets;

// James Ritchie Carroll - 2003


namespace TVA
{
    namespace Net
    {
        namespace Ftp
        {


            public class DataStream : Stream
            {



                private ControlChannel m_ctrl;
                private SessionConnected m_session;
                private System.Net.Sockets.TcpClient m_tcpClient;
                private Stream m_stream;
                private bool m_userAbort;

                internal DataStream(ControlChannel ctrl, System.Net.Sockets.TcpClient client)
                {

                    m_session = ctrl.Session;
                    m_ctrl = ctrl;
                    m_tcpClient = client;
                    m_stream = client.GetStream();
                    m_session.BeginDataTransfer(this);

                }

                public override void Close()
                {

                    if (!IsClosed)
                    {
                        CloseConnection();
                        m_ctrl.RefreshResponse();
                        m_ctrl.Session.EndDataTransfer();
                    }

                }

                internal void Abort()
                {

                    m_userAbort = true;

                }

                private void CloseConnection()
                {

                    m_stream.Close();
                    m_tcpClient.Close();
                    m_tcpClient = null;

                }

                internal bool IsClosed
                {
                    get
                    {
                        return m_tcpClient == null;
                    }
                }

                internal ControlChannel ControlChannel
                {
                    get
                    {
                        return m_ctrl;
                    }
                }

                public override bool CanRead
                {
                    get
                    {
                        return m_stream.CanRead;
                    }
                }

                public override bool CanSeek
                {
                    get
                    {
                        return m_stream.CanSeek;
                    }
                }

                public override bool CanWrite
                {
                    get
                    {
                        return m_stream.CanWrite;
                    }
                }

                public override long Length
                {
                    get
                    {
                        return m_stream.Length;
                    }
                }

                public override long Position
                {
                    get
                    {
                        return m_stream.Position;
                    }
                    set
                    {
                        m_stream.Position = value;
                    }
                }

                public override void Flush()
                {

                    m_stream.Flush();

                }

                public override long Seek(long offset, SeekOrigin origin)
                {

                    return m_stream.Seek(offset, origin);

                }

                public override int Read(byte[] buffer, int offset, int count)
                {

                    try
                    {
                        return m_stream.Read(buffer, offset, count);
                    }
                    finally
                    {
                        if (m_userAbort)
                        {
                            throw (new UserAbortException());
                        }
                    }

                }

                public override void Write(byte[] buffer, int offset, int count)
                {

                    try
                    {
                        m_stream.Write(buffer, offset, count);
                    }
                    finally
                    {
                        if (m_userAbort)
                        {
                            throw (new UserAbortException());
                        }
                    }

                }

                public override int ReadByte()
                {

                    try
                    {
                        return m_stream.ReadByte();
                    }
                    finally
                    {
                        if (m_userAbort)
                        {
                            throw (new UserAbortException());
                        }
                    }

                }

                public override void WriteByte(byte b)
                {

                    try
                    {
                        m_stream.WriteByte(b);
                    }
                    finally
                    {
                        if (m_userAbort)
                        {
                            throw (new UserAbortException());
                        }
                    }

                }

                public override void SetLength(long len)
                {

                    m_stream.SetLength(len);

                }

            }

            public class InputDataStream : DataStream
            {



                internal InputDataStream(ControlChannel ctrl, System.Net.Sockets.TcpClient client)
                    : base(ctrl, client)
                {


                }

                public override void Write(byte[] buffer, int offset, int count)
                {

                    throw (new NotSupportedException());

                }

                public override void WriteByte(byte b)
                {

                    throw (new NotSupportedException());

                }

                public override bool CanWrite
                {
                    get
                    {
                        return false;
                    }
                }

            }

            public class OutputDataStream : DataStream
            {



                internal OutputDataStream(ControlChannel ctrl, System.Net.Sockets.TcpClient client)
                    : base(ctrl, client)
                {


                }

                public override int Read(byte[] buffer, int offset, int count)
                {

                    throw (new NotSupportedException());

                }

                public override int ReadByte()
                {

                    throw (new NotSupportedException());

                }

                public override bool CanRead
                {
                    get
                    {
                        return false;
                    }
                }

            }

        }
    }
}

File: Source/OldCode/Archive/2004 TVA Code Library/Source/Ftp/DataStream.vb
' James Ritchie Carroll - 2003
Option Explicit On 

Imports System.IO
Imports System.Net.Sockets

Namespace Ftp

	Public Class DataStream

		Inherits Stream

		Private m_ctrl As ControlChannel
        Private m_session As SessionConnected
        Private m_tcpClient As TcpClient
        Private m_stream As Stream
        Private m_userAbort As Boolean

        Friend Sub New(ByVal ctrl As ControlChannel, ByVal client As TcpClient)

            m_session = ctrl.Session
            m_ctrl = ctrl
            m_tcpClient = client
            m_stream = client.GetStream()
            m_session.BeginDataTransfer(Me)

        End Sub

        Public Overrides Sub Close()

            If Not IsClosed Then
                CloseConnection()
                m_ctrl.RefreshResponse()
                m_ctrl.Session.EndDataTransfer()
            End If

        End Sub

        Friend Sub Abort()

            m_userAbort = True

        End Sub

        Private Sub CloseConnection()

            m_stream.Close()
            m_tcpClient.Close()
            m_tcpClient = Nothing

        End Sub

        Friend ReadOnly Property IsClosed() As Boolean
            Get
                Return m_tcpClient Is Nothing
            End Get
        End Property

        Friend ReadOnly Property ControlChannel() As ControlChannel
            Get
                Return m_ctrl
            End Get
        End Property

        Public Overrides ReadOnly Property CanRead() As Boolean
            Get
                Return m_stream.CanRead
            End Get
        End Property

        Public Overrides ReadOnly Property CanSeek() As Boolean
            Get
                Return m_stream.CanSeek
            End Get
        End Property

        Public Overrides ReadOnly Property CanWrite() As Boolean
            Get
                Return m_stream.CanWrite
            End Get
        End Property

        Public Overrides ReadOnly Property Length() As Long
            Get
                Return m_stream.Length
            End Get
        End Property

        Public Overrides Property Position() As Long
            Get
                Return m_stream.Position
            End Get
            Set(ByVal Value As Long)
                m_stream.Position = Value
            End Set
        End Property

        Public Overrides Sub Flush()

            m_stream.Flush()

        End Sub

        Public Overrides Function Seek(ByVal offset As Long, ByVal origin As SeekOrigin) As Long

            Return m_stream.Seek(offset, origin)

        End Function

        Public Overrides Function Read(ByVal buffer() As Byte, ByVal offset As Integer, ByVal count As Integer) As Integer

            Try
                Return m_stream.Read(buffer, offset, count)
            Finally
                If m_userAbort Then
                    Throw New UserAbortException()
                End If
            End Try

        End Function

        Public Overrides Sub Write(ByVal buffer() As Byte, ByVal offset As Integer, ByVal count As Integer)

            Try
                m_stream.Write(buffer, offset, count)
            Finally
                If m_userAbort Then
                    Throw New UserAbortException()
                End If
            End Try

        End Sub

        Public Overrides Function ReadByte() As Integer

            Try
                Return m_stream.ReadByte()
            Finally
                If m_userAbort Then
                    Throw New UserAbortException()
                End If
            End Try

        End Function

        Public Overrides Sub WriteByte(ByVal b As Byte)

            Try
                m_stream.WriteByte(b)
            Finally
                If m_userAbort Then
                    Throw New UserAbortException()
                End If
            End Try

        End Sub

        Public Overrides Sub SetLength(ByVal len As Long)

            m_stream.SetLength(len)

        End Sub

    End Class

	Public Class InputDataStream

		Inherits DataStream

		Friend Sub New(ByVal ctrl As ControlChannel, ByVal client As TcpClient)

			MyBase.New(ctrl, client)

		End Sub

		Public Overrides Sub Write(ByVal buffer() As Byte, ByVal offset As Integer, ByVal count As Integer)

			Throw New NotSupportedException()

		End Sub

		Public Overrides Sub WriteByte(ByVal b As Byte)

			Throw New NotSupportedException()

		End Sub

		Public Overrides ReadOnly Property CanWrite() As Boolean
			Get
				Return False
			End Get
		End Property

	End Class

	Public Class OutputDataStream

		Inherits DataStream

		Friend Sub New(ByVal ctrl As ControlChannel, ByVal client As TcpClient)

			MyBase.New(ctrl, client)

		End Sub

		Public Overrides Function Read(ByVal buffer() As Byte, ByVal offset As Integer, ByVal count As Integer) As Integer

			Throw New NotSupportedException()

		End Function

		Public Overrides Function ReadByte() As Integer

			Throw New NotSupportedException()

		End Function

		Public Overrides ReadOnly Property CanRead() As Boolean
			Get
				Return False
			End Get
		End Property

	End Class

End Namespace